Unlock failed cron jobs and set a high "last_checked" value to avoid continous re-check

MorrisJobke opened the pull request:
* fixes issue where cronjobs of a not-loaded app are marked as "still running" because there is a "reserved_at" value stored
* fixes #9992
* [x] add a repair step that sets the value of all jobs to 0 during upgrade to not run into this issue in the future for fast upgrades

MorrisJobke committed:
Unlock failed cron jobs and set a high "last_checked" value to avoid continous re-check

* fixes issue where cronjobs of a not-loaded app are marked as "still running" because there is a "reserved_at" value stored
* fixed #9992

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke committed:
Only wait for cron if fix for endless wait is included

* see https://github.com/nextcloud/server/issues/9992

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
I also added another check in the updater to only wait for cron if the upgrade is from a version after this commit to have the fixed cron jobs already included. It could still cause trouble if there is this cronjob entry, then an upgrade to 14.0.0.7 happens and then directly the next upgrade happens without cron.php to execute to clean up the reserved jobs. Maybe I will add another repair step to reset the "reserved_at" entries.

nickvergessen commented on lib/private/BackgroundJob/JobList.php:
```diff
@@ -214,6 +215,14 @@ public function getNext() {
 			$job = $this->buildJob($row);
 
 			if ($job === null) {
+				// set the last_checked to 12h in the future to not check failing jobs all over again
```
This is basically a non-exiting job or one of a disabled app, not a "failing job"

nickvergessen replied:
If `execute()` fails, it is still blocked for ever

jancborchardt reviewed (approved):
Fixes the update issue https://github.com/nextcloud/server/issues/9992 for me

tobiasKaminsky commented:
I ran into this, and with this PR, I had to manually set maintenance to false and reload the upgrade process.
Now it works fine :+1: 
(thanks @rullzer for pointing me here)

MorrisJobke committed:
Add repair job for pending cron jobs

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
Ready for review. I added a repair step and it worked over here.

rullzer reviewed (approved):
Does the trick for me.
I vote to merge as it at least improves the situation.

MorrisJobke committed:
Suppress potential SQL injection warning as they are false positives

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
I just updated the version again as master had it already and added two times:

```
* @suppress SqlInjectionChecker
```

because there were two false positives.
